lib/openstack_query/runners/runner_wrapper.py

```diff
@@ -34,80 +34,112 @@ def __init__(self, marker_prop_func: PropFunc, connection_cls=OpenstackConnectio
-                query_meta_params = {}
-                if kwargs:
-                    query_meta_params = self._parse_meta_params(conn, **kwargs)
-                resource_objects = self._run_query(
-                    conn, server_side_filters, **query_meta_params
+                resource_objects = self._run_with_openstacksdk(
+                    conn=conn, server_filters=server_side_filters, **kwargs
                 )
-                logger.info(
-                    "server-side query complete - time elapsed: %s seconds",
+                logger.debug(
+                    "server-side quer(y/ies) completed - found %s items. (time elapsed: %0.4f seconds)",
+                    len(resource_objects),
                     time.time() - start,
                 )
-                logger.debug("server-side query found %s items", len(resource_objects))
 
-        if client_side_filter_func and not server_side_filters:
+        if client_side_filters:
             logger.info("applying client side filters")
-            resource_objects = self._apply_client_side_filter(
-                resource_objects, client_side_filter_func
+            resource_objects = self._apply_client_side_filters(
+                items=resource_objects, filters=client_side_filters
             )
 
-        logger.info("found %s items in total", len(resource_objects))
+        logger.info(
+            "Query Complete! Found %s items. Time elapsed: %0.4f seconds",
+            len(resource_objects),
+            time.time() - start,
+        )
+        return resource_objects
+
+    def _run_with_openstacksdk(
+        self,
+        conn: OpenstackConnection,
+        server_filters: Optional[ServerSideFilters] = None,
+        **kwargs,
+    ):
+        if not server_filters:
+            server_filters = [None]
+
+        if kwargs:
+            kwargs = self._parse_meta_params(conn, **kwargs)
+
+        meta_param_log_str = "\n\t\t".join(
+            [f"{key}: '{val}'" for key, val in kwargs.items()]
+        )
+        logger.info("found %s set(s) of server-side filters", len(server_filters))
+        resource_objects = []
+        for i, query_filters in enumerate(server_filters, 1):
+            logger.debug("running query %s / %s", i, len(server_filters))
+
+            filters_log_str = "None (getting all)"
+            if query_filters:
+                filters_log_str = "\n\t".join(
+                    [f"{key}: '{val}'" for key, val in query_filters.items()]
+                )
+
+            logger.debug(
+                "Running OpenstackSDK Query with: "
+                "\n\tserver side filters: "
+                "\n%s "
+                "\n\n meta kwargs: "
+                "\n\t%s",
+                filters_log_str,
+                "None" if not meta_param_log_str else meta_param_log_str,
+            )
+
+            resource_objects.extend(self._run_query(conn, query_filters, **kwargs))
+
         return resource_objects
```

Comment:
IMO this should be in it's own class or the executor, as it bloats the runner with implementation details that can be abstracted / tested elsewhere

Reply:
future PR - I agree the wrapper is quite big - this is the case with a lot of components atm - I'll have a think about this